src/metabase/query_processor/middleware/prevent_infinite_recursive_preprocesses.clj

```diff
@@ -0,0 +1,22 @@
+(ns metabase.query-processor.middleware.prevent-infinite-recursive-preprocesses
+  (:require [clojure.tools.logging :as log]
+            [metabase.query-processor.error-type :as error-type]
+            [metabase.util.i18n :refer [tru]]))
+
+(def ^:private ^:dynamic *preprocessing-level* 1)
+
+(def ^:private ^:const max-preprocessing-level 20)
+
+(defn prevent-infinite-recursive-preprocesses
```

Comment:
this makes sense as a thing but was there something about this changeset specifically that required it?